lib/puppet/rest/routes.rb

```diff
@@ -8,10 +9,25 @@ def self.ca
       Route.new(api: '/puppet-ca/v1/', default_server: Puppet[:ca_server], default_port: Puppet[:ca_port])
     end
 
+    # Make an HTTP request to fetch the named certificate, using the given
+    # HTTP client.
+    # @param [Puppet::Rest::Client] client the HTTP client to use to make the request
+    # @param [String] name the name of the certificate to fetch
+    # @raise [Puppet::Rest::ResponseError] if the response status is not OK or
+    #        the response has an unknown content encoding
+    # @return [String] the PEM-encoded certificate or certificate bundle
     def self.get_certificate(client, name)
-      client.get("certificate/#{name}",
-                 header: { 'Accept' => 'text/plain',
-                           'accept-encoding' => ACCEPT_ENCODING })
+      header = { 'Accept' => 'text/plain', 'accept-encoding' => ACCEPT_ENCODING }
+      body = ''
+      client.get("certificate/#{name}", header: header) do  |response, chunk|
+        content_encoding = response.headers['Content-Encoding']
+        body << Puppet::Rest::Compression.uncompress_chunk(content_encoding, chunk)
+      end
```

Comment:
Can we create a local compressor instance and yield chunks to it so that we don't need to maintain global state:

```ruby
compression = Puppet::Rest::Compression.new
client.get("certificate/#{name}", header: header) do  |response, chunk|
  content_encoding = response.headers['Content-Encoding']
  compression.uncompress(content_encoding, chunk) do |uncompressed|
    body << uncompressed
  end
end
```

Probably best to explicitly close the compressor in an ensure block

Reply:
Updated.